New package DataFormats/L1TParticleFlow for Phase2 L1Trigger

rekovic opened the pull request:
#### PR description:

New package ```DataFormats/L1TParticleFlow``` for Phase2 L1Trigger.

These classes are intended to hold L1T objects reconstructed using PartcleFlow and PUPPI algorithms in back-end elecronics boards of Phase2 L1Trigger system:
- Layer1
- Layer2

as described in the Phase2 L1 TDR.

rekovic committed:
Add DataFormats/L1TParticleFlow.

Co-authored-by: Giovanni <gpetruc@gmail.com>
Co-authored-by: Philip <violatingcp@gmail.com>

rekovic committed:
Apply code-check.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14438

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @rekovic  for master.

It involves the following packages:

DataFormats/L1TParticleFlow


The following packages do not have a category, yet:

DataFormats/L1TParticleFlow
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@cmsbuild can you please review it and eventually sign? Thanks.
@rovere this is something you requested to watch as well.
@davidlange6, @silviodonato, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

rekovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7248a3278a73cea9c013e86fa341cf41197f4780
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5451/summary.html
CMSSW: CMSSW_11_1_X_2020-03-30-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5451/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 72
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691719
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

makortel commented on DataFormats/L1TParticleFlow/src/classes.h:
```diff
@@ -0,0 +1,42 @@
+    edm::RefVector<l1t::PFJetCollection> l1PFJetRefVector;
+    edm::RefVector<l1t::PFTauCollection> l1PFTauRefVector;
+  };
+}  // namespace DataFormats_L1TParticleFlow
```
This is not needed anymore.
```suggestion
```

rekovic replied:
done.

makortel commented on DataFormats/L1TParticleFlow/src/classes_def.xml:
```diff
@@ -0,0 +1,45 @@
+  <class name="l1t::PFCluster" ClassVersion="5">
+        <version ClassVersion="5" checksum="1668592434"/>
+        <version ClassVersion="4" checksum="387028836"/>
+        <version ClassVersion="3" checksum="2630618461"/>
```
The convention for new classes (that have never been stored) is to not include any development-time intermediate versions (i.e. start directly with class version 3).

rekovic replied:
done.

makortel commented on DataFormats/L1TParticleFlow/src/classes_def.xml:
```diff
@@ -0,0 +1,45 @@
+  <class name="edm::Ref<l1t::PFJetCollection>" />
+  <class name="edm::RefVector<l1t::PFJetCollection>" />
+
+  <class name="l1t::PFTau" />
```
Should this one have a class version as well?

rekovic replied:
Added.

rekovic committed:
Address comments.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14457

- This PR adds an extra 24KB to repository

rekovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #29347 was updated. @cmsbuild can you please check and sign again.

rekovic commented:
Thanks @makortel.

cmsbuild commented:
+1
Tested at: a347366879f3d4c5c62999cd7e38d08c746fc253
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5471/summary.html
CMSSW: CMSSW_11_1_X_2020-03-31-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5471/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 35
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691756
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

rekovic commented:
+1

rekovic commented:
assign l1

cmsbuild commented:
New categories assigned: l1

@benkrikler,@rekovic you have been requested to review this Pull request/Issue and eventually sign? Thanks

rekovic commented:
+1

rekovic commented:
assign upgrade

cmsbuild commented:
New categories assigned: upgrade

@kpedro88 you have been requested to review this Pull request/Issue and eventually sign? Thanks

kpedro88 commented on DataFormats/L1TParticleFlow/interface/PFCandidate.h:
```diff
@@ -0,0 +1,61 @@
+  class PFCandidate : public L1Candidate {
+  public:
+    typedef edm::Ptr<l1t::Muon> MuonRef;
+    enum Kind { ChargedHadron = 0, Electron = 1, NeutralHadron = 2, Photon = 3, Muon = 4 };
```
To fit with offline PF, better to name this `ParticleType`
(also the order is not the same as offline PF, but the order is also not consistent elsewhere in CMSSW)

kpedro88 commented on DataFormats/L1TParticleFlow/interface/PFCluster.h:
```diff
@@ -0,0 +1,76 @@
+    float ptError() const { return ptError_; }
+    void setPtError(float ptError) { ptError_ = ptError; }
+
+    bool isEM() const { return hwQual() == 1; }
```
simplify to `return hwQual();`

kpedro88 commented on DataFormats/L1TParticleFlow/interface/PFCluster.h:
```diff
@@ -0,0 +1,76 @@
+    void setPtError(float ptError) { ptError_ = ptError; }
+
+    bool isEM() const { return hwQual() == 1; }
+    void setIsEM(bool isEM) { setHwQual(isEM ? 1 : 0); }
```
simplify to `setHwQual(isEM);`

kpedro88 commented on DataFormats/L1TParticleFlow/interface/PFTau.h:
```diff
@@ -0,0 +1,45 @@
+    float chargedIso() const { return iso_; }
+    float fullIso() const { return fullIso_; }
+    int id() const { return id_; }
+    bool passLooseNN() const { return iso_ * (0.1 + 0.2 * (min(pt(), 100.))) * 1. / 20.1 > 0.05; }
```
lots of magic numbers here, should all be named constants

kpedro88 commented on DataFormats/L1TParticleFlow/interface/PFCluster.h:
```diff
@@ -0,0 +1,76 @@
+        : L1Candidate(PolarLorentzVector(pt, eta, phi, 0), hwpt, hweta, hwphi, /*hwQuality=*/isEM ? 1 : 0),
+          hOverE_(hOverE),
+          ptError_(ptError) {
+      setPdgId(isEM ? 22 : 130);
```
magic numbers should be named constants

kpedro88 commented on DataFormats/L1TParticleFlow/interface/PFCluster.h:
```diff
@@ -0,0 +1,76 @@
+    typedef std::vector<ConstituentAndFraction> ConstituentsAndFractions;
+
+    PFCluster() {}
+    PFCluster(float pt,
```
this constructor could call the other constructor (to reduce duplication)

rekovic replied:
These are two different constructors, on with PolarLorentzVector one with LorentzVector. So leaving as is.

kpedro88 commented on DataFormats/L1TParticleFlow/src/PFCandidate.cc:
```diff
@@ -0,0 +1,28 @@
+  setPdgIdFromKind_(charge, kind);
+}
+
+void l1t::PFCandidate::setPdgIdFromKind_(int charge, Kind kind) {
```
member functions typically don't end with underscore

silviodonato commented:
@smuzaffar @makortel @mrodozov could you assign the new package to l1and upgrade?

rekovic committed:
Address comments: rename enum Kind to ParticleType as in reco, simplify code, use constants to document cuts.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14523

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #29347 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 27599567a62c263625c3fffd376ba3deb9a80076
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5538/summary.html
CMSSW: CMSSW_11_1_X_2020-04-04-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5538/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 24 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 43
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691748
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

rekovic commented:
all comments addressed.

rekovic commented:
+1

rekovic commented:
@kpedro88 please review again or sign.

kpedro88 commented on DataFormats/L1TParticleFlow/interface/PFTau.h:
```diff
@@ -0,0 +1,56 @@
+    float fullIso() const { return fullIso_; }
+    int id() const { return id_; }
+    bool passLooseNN() const {
+      return iso_ * (0.1 + 0.2 * (min(pt(), PFTAU_NN_PT_CUTOFF))) * 1. / 20.1 > PFTAU_NN_LOOSE_CUT;
```
there are still magic numbers here: 0.1, 0.2, and `1./20.1` (the latter should especially be stored as a named constant, to avoid unnecessary repeated division operations)

rekovic committed:
Use named constants.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14541

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14541/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14541/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic committed:
Code format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14543

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #29347 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

rekovic commented:
+1

This should have the last comment addressed, and should pass all the test. 
Upon completion of tests, @kpedro88  Please review or sign.

cmsbuild commented:
-1

Tested at: 4fbe0b94346c60ac0c0270d4a37335fc71335d6f

CMSSW: CMSSW_11_1_X_2020-04-06-1700
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5558/summary.html

I found follow errors while testing this PR

**Failed tests:  AddOn**

* **AddOn**:

I found errors in the following addon tests:

cmsbuild commented:
Comparison job queued.

rekovic commented:
Did not expect the test to fail since the changes since last successful tests were trivial. 

The above report of failing  AddOn test strange as it does not point to any failing test workflows. The [report](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5558/addOnTests/) of AddOn tests and the overall [log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5558/addOnTests.log) file don't show any tests failing.  This could be a glitch. (attn @smuzaffar @silviodonato )

Will restart the tests.

rekovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5558/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 40 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 32
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691759
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

kpedro88 commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
-1

Tested at: 4fbe0b94346c60ac0c0270d4a37335fc71335d6f

CMSSW: CMSSW_11_1_X_2020-04-06-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5567/summary.html

I found follow errors while testing this PR

**Failed tests:  AddOn**

* **AddOn**:

I found errors in the following addon tests:

cmsbuild commented:
Comparison job queued.

rekovic commented:
@smuzaffar @silviodonato  Which AddOn test failed here ?

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5567/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 40 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 27
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691764
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

silviodonato commented on DataFormats/L1TParticleFlow/interface/PFCandidate.h:
```diff
@@ -0,0 +1,61 @@
+#include "DataFormats/L1TParticleFlow/interface/PFCluster.h"
+#include "DataFormats/L1TParticleFlow/interface/PFTrack.h"
+
+namespace l1t {
```
don't you want to use the same namespace as in https://github.com/cms-sw/cmssw/pull/29348/files#diff-4ad75913d7e9668f34e61ae8aa206c80R12 (ie. `l1tp2`)?

rekovic replied:
no.  l1tp2 used in case of CalorimeterPhase2 is to distinguish those Calo objects from the Phase1 Calo objects. 
In this case, there are no PF objects in Phase1, so "l1t" namespace suffices.

silviodonato commented on DataFormats/L1TParticleFlow/interface/PFCluster.h:
```diff
@@ -0,0 +1,76 @@
+#include "DataFormats/L1Trigger/interface/L1Candidate.h"
+#include "DataFormats/Common/interface/Ref.h"
+
+namespace l1t {
```
or `l1tp2` ?

rekovic replied:
no.

silviodonato commented on DataFormats/L1TParticleFlow/interface/PFJet.h:
```diff
@@ -0,0 +1,47 @@
+#include "DataFormats/L1Trigger/interface/L1Candidate.h"
+#include "DataFormats/Common/interface/Ptr.h"
+
+namespace l1t {
```
or `l1tp2` ?

rekovic replied:
no

silviodonato commented on DataFormats/L1TParticleFlow/interface/PFTau.h:
```diff
@@ -0,0 +1,63 @@
+#include <vector>
+#include "DataFormats/L1Trigger/interface/L1Candidate.h"
+
+namespace l1t {
```
or `l1tp2` ?

rekovic replied:
no

silviodonato commented on DataFormats/L1TParticleFlow/interface/PFTrack.h:
```diff
@@ -0,0 +1,87 @@
+#include "DataFormats/L1TrackTrigger/interface/TTTypes.h"
+#include "DataFormats/Common/interface/Ref.h"
+
+namespace l1t {
```
or `l1tp2` ?

rekovic replied:
no

silviodonato commented on DataFormats/L1TParticleFlow/BuildFile.xml:
```diff
@@ -0,0 +1,8 @@
+<use   name="DataFormats/L1Trigger"/>
+<use   name="DataFormats/L1TrackTrigger"/>
+<use   name="DataFormats/L1THGCal"/>
```
is `DataFormats/L1THGCal` actually used ?

rekovic replied:
can be removed.

silviodonato commented on DataFormats/L1TParticleFlow/interface/PFCandidate.h:
```diff
@@ -0,0 +1,61 @@
+    void setMuon(const MuonRef& ref) { muonRef_ = ref; }
+
+    /// PUPPI weight (-1 if not available)
+    float puppiWeight() const { return puppiWeight_; }
```
why you use a different approach for `puppiWeight` instead of something like
```
    const float& puppiWeight() const { return puppiWeight; }
    void setpuppiWeight(const float& val) { puppiWeight_ = puppiWeight; }
```
?

silviodonato replied:
well I see @makortel suggested the opposite here
https://github.com/cms-sw/cmssw/pull/29346#discussion_r400587009
Anyhow, please add setPuppiWeight, if you think it can be useful.

rekovic replied:
Because it is more efficient to return float by value.
Returning primitive types by reference is less efficient because it saves nothing and adds overhead.

rekovic replied:
`puppiWeight_ ` is set in constructor. 
  /// PUPPI weight (-1 if not available)
Not allowing change of weight through setter.

silviodonato commented on DataFormats/L1TParticleFlow/interface/PFCluster.h:
```diff
@@ -0,0 +1,76 @@
+      setPdgId(isEM ? 22 : 130);  // photon : non-photon(K0)
+    }
+
+    float hOverE() const { return hOverE_; }
```
what about `const float`?

rekovic replied:
You probably meant `const float &`.
Same as above, more efficient to return float type by value.

makortel replied:
Just to be clear, for value return types (i.e. `const float`) the `const` effectively has no meaning.

silviodonato reviewed (commented):
I'm not sure, but I see in 
https://cms-sw.github.io/cms_coding_rules.html#4----technical-coding-rules-1
4.2
Each header ﬁle contains one class declaration only. (*)
4.3
Header ﬁles must not contain any implementation except for class templates and code to be inlined.
while I see you are implementing some non-inline function.

silviodonato commented:
About the AddOn test failure, see #29415.

rekovic committed:
Remove unneeded include.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14575

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14575/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14575/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

makortel commented:
> I'm not sure, but I see in
> https://cms-sw.github.io/cms_coding_rules.html#4----technical-coding-rules-1
> 4.2
> Each header ﬁle contains one class declaration only. (*)
> 4.3
> Header ﬁles must not contain any implementation except for class templates and code to be inlined.
> while I see you are implementing some non-inline function.

They are all implicitly inline already (because of being fully defined inside class definitions).

rekovic committed:
Add override.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29347/14576

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #29347 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 0d65e8c7b6276c408ea29cac27e249426edc5768
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5592/summary.html
CMSSW: CMSSW_11_1_X_2020-04-07-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6eb1d/5592/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 35
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2691756
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

rekovic commented:
+1

kpedro88 commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
